### PR TITLE
fix: Command bar is broken

### LINKED ISSUE #1574: Command bar is broken
Shows "no result found" and impossible to navigate with up/down keys, even though there are some results

Screenshot 2023-09-14 at 13 19 25

When doing a fix, ideally write a test to make sure it doesn't happen again

### PR DESCRIPTION
**Pull request by rishi-raj-jain**
fix #1574

### CODE REVIEW

File: front/src/modules/command-menu/components/CommandMenu.tsx [hunk 1]
```diff
@@ -109,23 +109,37 @@ export const CommandMenu = () => {
...
-                  Icon={cmd.Icon}
-                  shortcuts={cmd.shortcuts || []}
-                  onClick={cmd.onCommandClick}
-                />
-              ))}
-          </StyledGroup>
-        )}
+        {!matchingCreateCommand &&
+          !matchingNavigateCommand &&
+          commandMenuCommands.filter((cmd) => cmd.type === CommandType.Create)
```
Comment (charlesBochet) [inline front/src/modules/command-menu/components/CommandMenu.tsx:114]: this looks to be the same check as machingCreateCommand, could you double check?

File: front/src/modules/command-menu/components/CommandMenu.tsx [hunk 2]
```diff
@@ -109,23 +109,37 @@ export const CommandMenu = () => {
...
+            .length < 1 &&
+          commandMenuCommands.filter(
```
Comment (charlesBochet) [inline front/src/modules/command-menu/components/CommandMenu.tsx:116]: this looks be the same purpose as matchingNavigateCommand, could you update matchingNavigateCommand to also check for the label and then remove the check on line 116 that will be duplicated with matchingNavigateCommand

File: front/src/modules/command-menu/components/CommandMenu.tsx [hunk 3]
```diff
@@ -197,25 +211,33 @@ export const CommandMenu = () => {
...
-                  key={cmd.shortcuts?.join('') ?? ''}
-                  to={cmd.to}
-                  label={cmd.label}
-                  shortcuts={cmd.shortcuts}
-                />
-              ))}
-          </StyledGroup>
-        )}
+        {!matchingNavigateCommand &&
+          commandMenuCommands.filter(
```
Comment (charlesBochet) [inline front/src/modules/command-menu/components/CommandMenu.tsx:215]: same here

File: front/src/modules/command-menu/components/CommandMenu.tsx [hunk 4]
```diff
@@ -109,23 +109,37 @@ export const CommandMenu = () => {
...
+            (cmd) =>
+              (cmd.shortcuts?.join('').includes(search?.toUpperCase()) ||
+                cmd.label?.toUpperCase().includes(search?.toUpperCase())) &&
+              cmd.type === CommandType.Navigate,
+          ).length < 1 &&
+          people.length < 1 &&
+          companies.length < 1 &&
+          activities.length < 1 && <StyledEmpty>No results found.</StyledEmpty>}
+        {!matchingCreateCommand &&
```
Comment (charlesBochet) [inline front/src/modules/command-menu/components/CommandMenu.tsx:125]: same here

**Review by charlesBochet — COMMENTED** [review]
@rishi-raj-jain Thanks a lot for the PR.
I've added comments.

Can you also update the storybook tests in `CommandMenu.stories.tsx` by adding additional stories to test the following cases:
- Default without search
- Search but matching at least 1 person, 1 company, 1 activty, 1 create command, 1 navigate command
- Search but not matching any person, company, activity but matching at least 1 create command, 1 navigate command
- search but not matching create commande, navigate command but matching at lest 1 person
- search not matching anything

To tests these cases, you might need to instantiate a `WrapperComponent` in your test that would:
call `addToCommandMenu` to add commands to the menu
return` <CommandMenu />` to display it.

Happy to provide more information if needed!

Also to run the tests:
- from `front` folder: `yarn storybook:dev` (browsable on localhost:6006)
- from another terminal in `front` folder: y`arn storybook:coverage`

Thanks a gain!

**Comment by charlesBochet** [comment]
@rishi-raj-jain Any update on this one? :)

**Comment by rishi-raj-jain** [comment]
@charlesBochet I'm on it, I'll be back in an hour or so

**Comment by rishi-raj-jain** [comment]
@charlesBochet I'm failing to understand the checks, can so to make it clear:

- we don't want to update the check `matchingCreateCommand`, correct? but in some instances I see 

```
{!matchingCreateCommand &&
          commandMenuCommands.filter((cmd) => cmd.type === CommandType.Create)
            .length > 0 && (
            <StyledGroup heading="Create">
              {commandMenuCommands
                .filter((cmd) => cmd.type === CommandType.Create)
                .map((cmd) => (
                  <CommandMenuItem
                    key={cmd.label}
                    to={cmd.to}
                    label={cmd.label}
                    Icon={cmd.Icon}
                    shortcuts={cmd.shortcuts || []}
                    onClick={cmd.onCommandClick}
                  />
                ))}
            </StyledGroup>
          )}
```

which is totally fine if that's expected, but we shall not show the `no results found` in cases like this. Hence I've put all the checks at the top for if nothing is present below, then only no results found is seen.

**Comment by rishi-raj-jain** [comment]
I do want to make it super clear that we could though do:

- have all checks in each matchCommand variable and only render them if they are present
- similarly with array items such as people, etc.
- if nothing of these is present, we show no results

Where I'm getting a blocker is what I see right now. There are mutliple states of not and if matchingCommand are found. What are the expectations here with those variables?

**Comment by rishi-raj-jain** [comment]
I do want to make it super clear that we could though do:

have all checks in each matchCommand variable and only render them if they are present
similarly with array items such as people, etc.
if nothing of these is present, we show no results
Where I'm getting a blocker is what I see right now. There are mutliple states of not and if matchingCommand are found. What are the expectations here with those variables?

**Comment by charlesBochet** [comment]
> I do want to make it super clear that we could though do:
> 
> * have all checks in each matchCommand variable and only render them if they are present
> * similarly with array items such as people, etc.
> * if nothing of these is present, we show no results
> 
> Where I'm getting a blocker is what I see right now. There are mutliple states of not and if matchingCommand are found. What are the expectations here with those variables?

Yes, I've taken a look at the current CommandMenu and I don't get it either, it looks completely broken.
here is my recommendation:
- we should have a matchingCreateCommand = commands array filtered on create + filtered by search keyword if provided
- same for matchingNavigateCommand
- same for matchingPeople
- same for matchingCompanies
- same for matchingActivities
- if (!matchingCreateCommand.length && !matchingNavigateCommand && ... && !matchingActivities.length) ==> display No results found

**Comment by rishi-raj-jain** [comment]
@charlesBochet 

Awesome, I'll be on this in about an hour. Appreciate the detailed responses!

**Comment by rishi-raj-jain** [comment]
@charlesBochet this is ready for you to review.

**Comment by rishi-raj-jain** [comment]
@charlesBochet this is ready for you to review.

File: front/src/modules/command-menu/components/WrapperCommandMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,248 @@
+import { useState } from 'react';
```
Comment (charlesBochet) [inline front/src/modules/command-menu/components/WrapperCommandMenu.tsx:1]: Very nice, let's move this component within __stories__ folder as it's used for testing

**Review by charlesBochet — APPROVED** [review]

**Comment by rishi-raj-jain** [comment]
@charlesBochet done!

**Comment by charlesBochet** [comment]
Thinking about it, we should not introduce this Wrapper component for tests and use mocks instead! I'm merging it as it is and I'm refactoring the tests

**Comment by rishi-raj-jain** [comment]
Awesome! ✨

Can I get paid for the same on Algora?

**Pull request merged**